system/lib/pthread/emscripten_yield.c

```diff
@@ -28,7 +28,11 @@ void _emscripten_yield(double now) {
   // allocate (or otherwise itself crash) so use a low level atomic primitive
   // for this signal.
   if (is_runtime_thread) {
-    if (thread_crashed) {
+    if (crashed_thread_id) {
+      // Mark the crashed thread as strongly referenced so that Node.js doesn't
+      // exit while the pthread is propagating the uncaught exception back to
+      // the main thread.
+      _emscripten_thread_set_strongref(crashed_thread_id);
```

Comment:
I actually think this change might help with the race conditions we've been seeing in certain pthread tests.

Reply:
Is is possible that by the time we get here, the `crashed_thread_id` has already be cleaned up and that `crashed_thread_id` is no longer a valid thread pointer?

Reply:
> I actually think this change might help with the race conditions we've been seeing in certain pthread tests.

Did you mean issue #15014? I think that race condition is still there, I just responded to that issue.

> Is is possible that by the time we get here, the `crashed_thread_id` has already be cleaned up and that `crashed_thread_id` is no longer a valid thread pointer?

Given that crashed threads are never cleaned up (i.e. they do not call `_emscripten_thread_exit`), I think we can be sure that the pthread and corresponding worker are still alive up once it arrives here.

There could be a possibility that the uncaught exception will be re-thrown on the main thread before it ever arrives here, but since the program will be finished by then, I don't think that would be an issue.